src/Util/FisherYates/Implementation.dfy

```diff
@@ -0,0 +1,33 @@
+/*******************************************************************************
+ *  Copyright by the contributors to the Dafny Project
+ *  SPDX-License-Identifier: MIT
+ *******************************************************************************/
+
+module FisherYates.Implementation {
+  import Interface
+
+  trait {:termination false} Trait extends Interface.Trait {
+
+    method Shuffle<T(!new)>(a: array<T>)
+      decreases *
+      modifies this, a
+    {
+      var i := 0;
```

Comment:
Why not a for loop?

Reply:
I don't recall the details anymore, but it had to do with the equivalence proof. I'll turn it into a for loop, and change it back into a while loop later on, if really necessary.